### PR TITLE
Added warning for TLS validation

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
Please provide a description for what this PR is for.

And to help us figure out who should review this PR, please 
put an X in all the areas that this PR affects.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any changes that may affect Istio users.

### CODE REVIEW

**Review by howardjohn — COMMENTED** [review]
Are these currently accepted today? Restricting validation is a breaking API change if not so I am not sure this is safe. In particular, caCertificates seems like it's legitimate to not specify.

**Comment by hzxuzhonghu** [comment]
ISTIO_MUTUAL mode should not manually provide key/cert, it seems this is missing since the beginning.

**Review by howardjohn — CHANGES_REQUESTED** [review]
The issue is not about whether the configuration is correct or not - if we allowed the configuration in the past we *must* allow it in the future. This is a core requirement of making Istio upgrades backwards compatible. I have data to suggest there are a large number of DestinationRules that would fail this validation check in real Istio users clusters, so I don't think this is safe.

**Comment by howardjohn** [comment]
We can add warnings but not an error

**Comment by duderino** [comment]
@hzxuzhonghu backwards compatibility breaks need to be approved by the TOC.  I understand why you want this PR (lack of validation here is a real weakness), but John is right that we have to be very careful about preserving backwards compatibility going forward.

Another option is to add another warning to https://istio.io/latest/docs/ops/best-practices/security/.

@howardjohn @hzxuzhonghu I wonder if we can add an opt-in flag like 'strict-security'?  When that flag is on, failed validation checks like this would be errors.  When the flag is off, failed validation checks like this would be warnings.  Thoughts?

**Comment by howardjohn** [comment]
I don't think its a security issue, it's just not the "right" thing to do.
I think just a warning in the webhook would be good. then when a user
applies the config they immediately get a warning. Istioctl analyze will
also show it.

**Comment by hzxuzhonghu** [comment]
I agree it is safer to warn first and then we can prevent it maybe in two releases later v1.12.   
We have to admit defects that have long existed, lol.

**Comment by howardjohn** [comment]
Making validation stricter is the same as making a breaking changes to the
api like removing a field - we don't want to prevent people from upgrading.

**Comment by hzxuzhonghu** [comment]
@howardjohn Changed to warning to not break upgrade.

File: pkg/config/validation/validation.go [hunk 1]
```diff
@@ -1129,33 +1129,45 @@ func validateLoadBalancer(settings *networking.LoadBalancerSettings) (errs error
...
 		if settings.ClientCertificate == "" {
-			errs = appendErrors(errs, fmt.Errorf("client certificate required for mutual tls"))
+			v = appendValidation(v, fmt.Errorf("client certificate required for mutual tls"))
 		}
 		if settings.PrivateKey == "" {
-			errs = appendErrors(errs, fmt.Errorf("private key required for mutual tls"))
+			v = appendValidation(v, fmt.Errorf("private key required for mutual tls"))
+		}
+		if settings.CaCertificates == "" {
+			v = appendValidation(v, WrapWarning(fmt.Errorf("ca certificate required for mutual tls,  it will fail in release 1.12")))
```
Comment (howardjohn) [inline pkg/config/validation/validation.go:1158]: Is it actually required? or does not setting it allow us to have optional verification? ie equivalent to golang `tls.RequestClientCert`. https://github.com/istio/istio/issues/31984

File: pkg/config/validation/validation.go [hunk 2]
```diff
@@ -1129,33 +1129,45 @@ func validateLoadBalancer(settings *networking.LoadBalancerSettings) (errs error
...
 		}
 	}
 
-	return
+	if settings.Mode == networking.ClientTLSSettings_ISTIO_MUTUAL {
+		if settings.CredentialName != "" || settings.ClientCertificate != "" ||
+			settings.CaCertificates != "" || settings.PrivateKey != "" {
+			v = appendValidation(v, WrapWarning(fmt.Errorf("cannot specify privateKey, clientCertificate, caCertificates or"+
+				" credentialName for ISTIO_MUTUAL tls, it will fail in release 1.12")))
```
Comment (howardjohn) [inline pkg/config/validation/validation.go:1166]:
```suggestion
				" credentialName for ISTIO_MUTUAL tls")))
```

File: pkg/config/validation/validation.go [hunk 3]
```diff
@@ -1469,9 +1481,14 @@ func ValidateProxyConfig(config *meshconfig.ProxyConfig) (errs error) {
...
 
 	if tracer := config.GetTracing().GetTlsSettings(); tracer != nil {
-		if err := validateTLS(tracer); err != nil {
+		v := validateTLS(tracer)
+		err, warning := v.Unwrap()
+		if err != nil {
 			errs = multierror.Append(errs, multierror.Prefix(err, "invalid tracing TLS config:"))
 		}
+		if warning != nil {
+			scope.Warnf("invalid tracing TLS config: %v", warning)
```
Comment (howardjohn) [inline pkg/config/validation/validation.go:1490]: Why are we logging rather than returning the warning?

Comment (hzxuzhonghu) [reply pkg/config/validation/validation.go:1490]: This is in meshconfig, which is not validated by webhook

File: pkg/config/validation/validation.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pkg/config/validation/validation.go:1158]: make sense

Comment (hzxuzhonghu) [reply pkg/config/validation/validation.go:1158]: Nope, this is mutual tls, not ClientTLSSettings_SIMPLE

File: pkg/config/validation/validation.go (hunk 2 above)
Comment (hzxuzhonghu) [reply pkg/config/validation/validation.go:1166]: don't we need to give this note? Then we can strictly validate it.

File: pkg/config/validation/validation.go (hunk 1 above)
Comment (howardjohn) [reply pkg/config/validation/validation.go:1158]: Yes I know. You can have mutual TLS without caCert (in theory - I don't know if it works in Istio today). See the linked issue

Comment (howardjohn) [reply pkg/config/validation/validation.go:1158]: basically optional mtls

File: pkg/config/validation/validation.go (hunk 2 above)
Comment (howardjohn) [reply pkg/config/validation/validation.go:1166]: As mentioned above, we cannot ever strictly validate it. Its a breaking API change.

Comment (hzxuzhonghu) [reply pkg/config/validation/validation.go:1166]: I would suggest doing this like removing an API field or a whole API

**Pull request closed without merging**